WIP: Add KEP for PersistentVolume release timestamp (#19)
jsafrane/enhancements: keps/sig-storage/3762-persistent-volume-last-phase-transition-time/README.md

useful for a wide audience.

A good summary is probably at least a paragraph in length.

Both in this section and below, follow the guidelines of the [documentation
style guide]. In particular, wrap lines to a reasonable length, to make it
easier for reviewers to cite specific portions, and to minimize diff churn on
updates.

[documentation style guide]: https://github.com/kubernetes/community/blob/master/contributors/guide/style-guide.md
-->

We want to add a new PersistentVolumeStatus field, which would hold a timestamp of when a PersistentVolume last
transitioned to a different phase.

## Motivation

Some users have experienced data loss when using `Delete` retain policy and reverted to a safer `Retain` policy.
With `Retain` policy all volumes that are retained and left unclaimed have their phase is set to `Released`.
As the released volumes pile up over time admins want to perform manual cleanup based on the time when the volume was
last used, which is when the volume transitioned to `Released` phase.

We can approach the solution in a more generic way and record a timestamp of when the volume transitioned to any phase,

> not just to `Released` phase. This is more informative for admins as it covers more scenarios while it still meets the
> original expectation described above.

Reviewer: ```suggestion
not just to `Released` phase. This allows anyone, incl. our perf tests, to measure time e.g. between a PV `Pending` and `Bound`
```

<!--
This section is for explicitly listing the motivation, goals, and non-goals of
this KEP.  Describe why the change is important and the benefits to users. The
motivation section can optionally provide links to [experience reports] to
demonstrate the interest in a KEP within the wider Kubernetes community.

[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
-->

### Goals

1) Introduce a new status field in PersistentVolumes.

> 2) Update the new field with a timestamp every time a volume transitions to a different phase.

Reviewer: ```suggestion
2) Update the new field with a timestamp every time a volume transitions to a different phase (`pv.Status.Phase`).
```

This should have enough detail that reviewers can understand exactly what
you're proposing, but should not include things like API designs or
implementation. What is the desired outcome and how do we measure success?.
The "Design Details" section below is for the real
nitty-gritty.
-->

We need to update API server to support the newly proposed field and update PV controller to set a value of the new
timestamp field when a volume transitions to a different phase. Also, if the feature gate is disabled the value must be
re-set to `nil` when updating or creating a volume.

The value of the field is not intended for use by any other Kubernetes components at this point and should be used only
as a convenience feature for cluster admins. Cluster admins should be able to list and sort PersistentVolumes based on
a timestamp which indicates when the volume transitioned to a different state. 

### User Stories (Optional)

<!--
Detail the things that people will be able to do if this KEP is implemented.
Include as much detail as possible so that people can understand the "how" of
the system. The goal here is to make this feel real for users without getting
bogged down.
-->

#### Story 1

> As a cluster admin I want to use `Retain` policy for released volumes which is safer than `Delete` and gives me more
> control over volume cleanup. However, this in turn leads to a growing number of volumes that need to be monitored
> and cleaned up. The cleanup should be performed based how long the volume was not used by any claim. I should be able
> to list all volumes that are now in `Released` phase and see when it transitioned to this phase to decide which volumes
> can be safely removed.

Reviewer: ```suggestion
As a cluster admin I want to use `Retain` policy for released volumes, which is safer than `Delete`, and implement a reliable policy to delete volumes that are `Released` for more than X days.
```

### Notes/Constraints/Caveats (Optional)

<!--
What are the caveats to the proposal?
What are some important details that didn't come across above?
Go in to as much detail as necessary here.
This might be a good place to talk about core concepts and how they relate.
-->

### Risks and Mitigations

<!--
What are the risks of this proposal, and how do we mitigate? Think broadly.
For example, consider both security and how this will impact the larger
Kubernetes ecosystem.

How will security be reviewed, and by whom?

How will UX be reviewed, and by whom?

Consider including folks who also work outside the SIG or subproject.
-->

The new field is purely informative and should not introduce any risk.

## Design Details

Changes required for this KEP:

* kube-apiserver
  * extend PersistentVolumeStatus type with `LastPhaseTransitionTime` field
  * validation should check timestamp format 
  * validation should allow update from `nil`
  * validation should not allow updating to a point in time before the current timestamp

* kube-controller-manager / PV controller

>   * update the timestamp whenever PV controller transitions PV to a different phase

Reviewer: ```suggestion
  * update the timestamp whenever PV controller transitions PV to a different phase (`pv.Status.Phase`)
```

  * remove the timestamp in `LastPhaseTransitionTime` field during volume status update when feature gate is disabled

>     * extend and use the existing function for disabling a status field: https://github.com/kubernetes/kubernetes/blob/d94261e904c90578e913c42c4d2a0fc8cb30937f/pkg/api/persistentvolumeclaim/util.go#L91

Reviewer: I think util.go is used in the API server in PrepareForUpdate / Create. Do you really want the controller to clear the field proactively?

Author: Removing this from KEP for now - it does not have to be that specific probably, I'll figure out the right place later.

for the KEP to move to `implementable` status and be included in the release.

In some cases, the questions below should also have answers in `kep.yaml`. This
is to enable automation to verify the presence of the review, and to reduce review
burden and latency.

The KEP must have a approver from the
[`prod-readiness-approvers`](http://git.k8s.io/enhancements/OWNERS_ALIASES)
team. Please reach out on the
[#prod-readiness](https://kubernetes.slack.com/archives/CPNHUMN74) channel if
you need any help or guidance.
-->

### Feature Enablement and Rollback

<!--
This section must be completed when targeting alpha to a release.
-->

###### How can this feature be enabled / disabled in a live cluster?

<!--
Pick one of these and delete the rest.

Documentation is available on [feature gate lifecycle] and expectations, as
well as the [existing list] of feature gates.

[feature gate lifecycle]: https://git.k8s.io/community/contributors/devel/sig-architecture/feature-gates.md
[existing list]: https://kubernetes.io/docs/reference/command-line-tools-reference/feature-gates/
-->

- [X] Feature gate (also fill in values in `kep.yaml`)
  - Feature gate name:

>   - Components depending on the feature gate:

Reviewer: File this please

Author: Done.
